server/src/main/java/org/elasticsearch/search/internal/LegacyReaderContext.java

```diff
@@ -61,6 +61,7 @@ public LegacyReaderContext(long id, IndexService indexService, IndexShard indexS
             if (searcher == null) {
                 Engine.Searcher delegate = searcherSupplier.acquireSearcher(source);
                 onClose = delegate::close;
+                // wrap the searcher so that closing is a noop, the actual closing happens when this context is closed
```

Comment:
@jimczi I looked at this again (while investigating the perf issue). I think we need to double-check with synchronization before acquiring a searcher to avoid leaking it.

Reply:
Or we can just always `addOnClose(delegate)` instead.

Reply:
I found another issue while investigating the perf issue. We don't have a happens-before relation between the initial search and subsequent scroll requests. The `searcher` field needs to be `volatile`; otherwise, we might acquire and wrap searcher more than once.  I've pushed https://github.com/elastic/elasticsearch/pull/61831/commits/0c42f9c0b1740cb75d511118dca90459d9bdc38c. Let me know if you are ok with this change.

Reply:
Great catch, thanks!